src/librustc_middle/ty/sty.rs

```diff
@@ -395,16 +418,49 @@ pub struct GeneratorSubsts<'tcx> {
     pub substs: SubstsRef<'tcx>,
 }
 
-struct SplitGeneratorSubsts<'tcx> {
-    resume_ty: GenericArg<'tcx>,
-    yield_ty: GenericArg<'tcx>,
-    return_ty: GenericArg<'tcx>,
-    witness: GenericArg<'tcx>,
-    tupled_upvars_ty: GenericArg<'tcx>,
+// FIXME(eddyb) find a better name than `Split...` - perhaps `GeneratorSubstsParts`?
+pub struct SplitGeneratorSubsts<T> {
+    pub resume_ty: T,
+    pub yield_ty: T,
+    pub return_ty: T,
+    pub witness: T,
+    pub tupled_upvars_ty: T,
 }
 
 impl<'tcx> GeneratorSubsts<'tcx> {
-    fn split(self) -> SplitGeneratorSubsts<'tcx> {
+    /// Construct `GeneratorSubsts` from base `Substs` (i.e. of the parent function)
+    /// and `SplitGeneratorSubsts` containing the additional generator-specific components.
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        base_substs: SubstsRef<'tcx>,
+        parts: SplitGeneratorSubsts<Ty<'tcx>>,
+    ) -> GeneratorSubsts<'tcx> {
+        GeneratorSubsts {
+            substs: tcx.mk_substs(
+                base_substs.iter().chain(
+                    [
+                        parts.resume_ty,
+                        parts.yield_ty,
+                        parts.return_ty,
+                        parts.witness,
+                        parts.tupled_upvars_ty,
+                    ]
+                    .iter()
+                    .map(|&ty| ty.into()),
+                ),
+            ),
+        }
+    }
+
+    /// Returns the non-generator-specific base `Substs` (i.e. of the parent function).
+    /// The ordering assumed here must match that used by `GeneratorSubsts::new` above.
+    pub fn base_substs(self) -> &'tcx [GenericArg<'tcx>] {
+        &self.substs[..self.substs.len() - 5]
```

Comment:
Same here should this assert(is_valid)? [`as_closure`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/subst.rs.html#193)